### ISSUE TITLE
HTTPS probes with resolve_first ignore tls_config.server_name

### ISSUE THREAD

**Issue by cbroglie** [cloudprober/cloudprober#294]
**Describe the bug**
Given a config like this, the `server_name` will be ignored:
```
http_probe {
  headers {
    name: "Host"
    value: "example.com"
  }
  protocol: HTTPS
  relative_url: "/ping"
  resolve_first: true
  tls_config {
    server_name: "example.com"
  }
}
```

This is b/c #159 started setting `TLSClientConfig.ServerName` unconditionally, overwriting any existing value set in `baseTransport`.

This patch fixes the issue, and I can follow up with a PR:
```diff
diff --git a/probes/http/http.go b/probes/http/http.go
index 99dde7c..076444c 100644
--- a/probes/http/http.go
+++ b/probes/http/http.go
@@ -468,7 +468,9 @@ func (p *Probe) clientsForTarget(target endpoint.Endpoint) []*http.Client {
 				if t.TLSClientConfig == nil {
 					t.TLSClientConfig = &tls.Config{}
 				}
-				t.TLSClientConfig.ServerName = urlHostForTarget(target)
+				if t.TLSClientConfig.ServerName == "" {
+					t.TLSClientConfig.ServerName = urlHostForTarget(target)
+				}
 			}
 			clients[i] = &http.Client{Transport: t}
 		} else {

```

**Cloudprober Version**
v0.12.3

**Comment by manugarg** [comment]
Indeed. Thank you so much for catching and reporting this, @cbroglie. Your proposal LGTM.

**Comment by manugarg** [comment]
@cbroglie, since I cut a new release today itself, I'll just go ahead cut another release to include this bug fix and deprecate v0.12.4.

**Comment by cbroglie** [comment]
Appreciate the quick turnaround!
